From bdeb68730843294204052795179edc5dd784eb09 Mon Sep 17 00:00:00 2001
From: John Galt <johngaltfirstrun@gmail.com>
Date: Sat, 20 Jun 2026 13:33:27 +0800
Subject: [PATCH] accessibility/vibration: Introduce Haptic Effect Style
 selection

We use debug.haptic_profile and device init side to set vendor persist
property so settings can set the property.
We also use a settings key since debug props are not persistent.

[1] https://github.com/yaap/device_oneplus_sm8650-common/commit/ca25698af3470402639d11d62a600c52bdf0e66b

Change-Id: I47659a58406f36de4ce2d26f0bc4cc39536109dc
---
 res/values/config.xml                         |  3 +
 res/values/matrixx_arrays.xml                 | 13 +++
 res/values/matrixx_strings.xml                |  3 +
 ...ssibility_vibration_intensity_settings.xml |  7 ++
 res/xml/accessibility_vibration_settings.xml  |  7 ++
 .../HapticStyleListPreference.java            | 23 ++++++
 .../HapticStylePreferenceController.java      | 79 +++++++++++++++++++
 7 files changed, 135 insertions(+)
 create mode 100644 src/com/android/settings/accessibility/HapticStyleListPreference.java
 create mode 100644 src/com/android/settings/accessibility/HapticStylePreferenceController.java

diff --git a/res/values/config.xml b/res/values/config.xml
index 131465a37a8..51c39fe762c 100644
--- a/res/values/config.xml
+++ b/res/values/config.xml
@@ -566,6 +566,9 @@
     <string-array name="config_force_battery_unrestrict_mode_apps" translatable="false">
     </string-array>
 
+    <!-- Show haptic styles -->
+    <bool name="config_supportHapticProfiles" translatable="false">true</bool>
+
     <!-- Uri to query non-public Slice Uris. -->
     <string name="config_non_public_slice_query_uri" translatable="false"></string>
 
diff --git a/res/values/matrixx_arrays.xml b/res/values/matrixx_arrays.xml
index 719b3ade1b1..7b473b48e96 100644
--- a/res/values/matrixx_arrays.xml
+++ b/res/values/matrixx_arrays.xml
@@ -83,4 +83,17 @@
         <item>1</item>
         <item>0</item>
     </string-array>
+
+    <!-- Haptic Effect Styles -->
+    <string-array name="haptic_effect_style_entries" translatable="false">
+        <item>RichTap</item>
+        <item>Crisp</item>
+        <item>Gentle</item>
+    </string-array>
+
+    <string-array name="haptic_effect_style_values" translatable="false">
+        <item>richtap</item>
+        <item>crisp</item>
+        <item>gentle</item>
+    </string-array>
 </resources>
diff --git a/res/values/matrixx_strings.xml b/res/values/matrixx_strings.xml
index afe69698540..a7b610243af 100644
--- a/res/values/matrixx_strings.xml
+++ b/res/values/matrixx_strings.xml
@@ -131,6 +131,9 @@
     <string name="back_gesture_haptic_title">Back gesture haptic</string>
     <string name="back_gesture_haptic_summary">Vibrate when performing a back gesture</string>
 
+    <!-- Haptic Effect Style -->
+    <string name="haptic_effect_style_title">Haptic effect style</string>
+
     <!-- Gesture navbar radius -->
     <string name="gesture_navbar_radius_title">Pill radius</string>
     <string name="thin_label">Thin</string>
diff --git a/res/xml/accessibility_vibration_intensity_settings.xml b/res/xml/accessibility_vibration_intensity_settings.xml
index 451c1dffa95..946044abff5 100644
--- a/res/xml/accessibility_vibration_intensity_settings.xml
+++ b/res/xml/accessibility_vibration_intensity_settings.xml
@@ -65,6 +65,13 @@
         android:key="vibration_category_haptics"
         android:title="@string/accessibility_interactive_haptics_category_title">
 
+        <com.android.settings.accessibility.HapticStyleListPreference
+            android:key="haptic_style"
+            android:title="@string/haptic_effect_style_title"
+            android:entries="@array/haptic_effect_style_entries"
+            android:entryValues="@array/haptic_effect_style_values"
+            app:controller="com.android.settings.accessibility.HapticStylePreferenceController" />
+
         <com.android.settingslib.widget.SliderPreference
             android:key="haptic_feedback_intensity"
             android:title="@string/accessibility_touch_vibration_title"
diff --git a/res/xml/accessibility_vibration_settings.xml b/res/xml/accessibility_vibration_settings.xml
index 41182703de0..15856af7bb8 100644
--- a/res/xml/accessibility_vibration_settings.xml
+++ b/res/xml/accessibility_vibration_settings.xml
@@ -68,6 +68,13 @@
         android:title="@string/accessibility_interactive_haptics_category_title"
         android:persistent="false">
 
+        <com.android.settings.accessibility.HapticStyleListPreference
+            android:key="haptic_style"
+            android:title="@string/haptic_effect_style_title"
+            android:entries="@array/haptic_effect_style_entries"
+            android:entryValues="@array/haptic_effect_style_values"
+            app:controller="com.android.settings.accessibility.HapticStylePreferenceController" />
+
         <SwitchPreferenceCompat
             android:key="toggle_haptic_feedback_intensity"
             android:title="@string/accessibility_touch_vibration_title"
diff --git a/src/com/android/settings/accessibility/HapticStyleListPreference.java b/src/com/android/settings/accessibility/HapticStyleListPreference.java
new file mode 100644
index 00000000000..f89f6826528
--- /dev/null
+++ b/src/com/android/settings/accessibility/HapticStyleListPreference.java
@@ -0,0 +1,23 @@
+package com.android.settings.accessibility;
+
+import android.content.Context;
+import android.os.SystemProperties;
+import android.provider.Settings;
+import android.util.AttributeSet;
+import androidx.preference.ListPreference;
+
+public class HapticStyleListPreference extends ListPreference {
+    private static final String PROP_KEY = "persist.sys.haptic_profile";
+    private static final String SETTINGS_KEY = Settings.Secure.HAPTIC_EFFECTS_PROFILE;
+
+    public HapticStyleListPreference(Context context, AttributeSet attrs) {
+        super(context, attrs);
+    }
+
+    @Override
+    public void setValue(String value) {
+        super.setValue(value);
+        SystemProperties.set(PROP_KEY, value);
+        Settings.Secure.putString(getContext().getContentResolver(), SETTINGS_KEY, value);
+    }
+}
diff --git a/src/com/android/settings/accessibility/HapticStylePreferenceController.java b/src/com/android/settings/accessibility/HapticStylePreferenceController.java
new file mode 100644
index 00000000000..a5eb8ca1ac3
--- /dev/null
+++ b/src/com/android/settings/accessibility/HapticStylePreferenceController.java
@@ -0,0 +1,79 @@
+package com.android.settings.accessibility;
+
+import android.content.Context;
+import android.os.SystemProperties;
+import android.provider.Settings;
+import androidx.preference.ListPreference;
+import androidx.preference.Preference;
+import androidx.preference.PreferenceScreen;
+
+import com.android.settings.core.BasePreferenceController;
+import com.android.settings.R;
+
+public class HapticStylePreferenceController extends BasePreferenceController
+        implements Preference.OnPreferenceChangeListener {
+
+    private static final String PROP_KEY = "debug.haptic_profile";
+    private static final String SETTINGS_KEY = Settings.Secure.HAPTIC_EFFECTS_PROFILE;
+    private ListPreference mPreference;
+
+    public HapticStylePreferenceController(Context context, String preferenceKey) {
+        super(context, preferenceKey);
+    }
+
+    @Override
+    public int getAvailabilityStatus() {
+        return mContext.getResources().getBoolean(R.bool.config_supportHapticProfiles) ? AVAILABLE : UNSUPPORTED_ON_DEVICE;
+    }
+
+    @Override
+    public void displayPreference(PreferenceScreen screen) {
+        super.displayPreference(screen);
+        Preference pref = screen.findPreference(getPreferenceKey());
+        if (pref instanceof ListPreference) {
+            mPreference = (ListPreference) pref;
+            mPreference.setOnPreferenceChangeListener(this);
+            updateState(mPreference);
+        }
+    }
+
+    @Override
+    public void updateState(Preference preference) {
+        super.updateState(preference);
+        if (preference instanceof ListPreference) {
+            ListPreference listPref = (ListPreference) preference;
+            String currentValue = Settings.Secure.getString(mContext.getContentResolver(), SETTINGS_KEY);
+
+            // If the property is empty, use the first entry value as default
+            if (currentValue == null || currentValue.isEmpty()) {
+                CharSequence[] entryValues = listPref.getEntryValues();
+                if (entryValues != null && entryValues.length > 0) {
+                    currentValue = entryValues[0].toString();
+                    listPref.setValue(currentValue);
+                    SystemProperties.set(PROP_KEY, currentValue);
+                    Settings.Secure.putString(mContext.getContentResolver(), SETTINGS_KEY, currentValue);
+                }
+            } else {
+                listPref.setValue(currentValue);
+            }
+
+            int index = listPref.findIndexOfValue(currentValue);
+            if (index >= 0) {
+                preference.setSummary(listPref.getEntries()[index]);
+            }
+        }
+    }
+
+    @Override
+    public boolean onPreferenceChange(Preference preference, Object newValue) {
+        if (!(preference instanceof ListPreference)) return false;
+        String value = (String) newValue;
+        SystemProperties.set(PROP_KEY, value);
+        Settings.Secure.putString(mContext.getContentResolver(), SETTINGS_KEY, value);
+        int index = mPreference.findIndexOfValue(value);
+        if (index >= 0) {
+            mPreference.setSummary(mPreference.getEntries()[index]);
+        }
+        return true;
+    }
+}
-- 
2.54.0

